Assignment/PingProcess.cs

```diff
@@ -1,173 +1,174 @@
-            {
-                process.Kill();
-            }
-
-        }
-        return process;
-
-        void OutputHandler(object s, DataReceivedEventArgs e)
-        {
-            progressOutput?.Invoke(e.Data);
-        }
-
-        void ErrorHandler(object s, DataReceivedEventArgs e)
-        {
-            progressError?.Invoke(e.Data);
-        }
-    }
-
-    private static ProcessStartInfo UpdateProcessStartInfo(ProcessStartInfo startInfo)
-    {
-        startInfo.CreateNoWindow = true;
-        startInfo.RedirectStandardError = true;
-        startInfo.RedirectStandardOutput = true;
-        startInfo.UseShellExecute = false;
-        startInfo.WindowStyle = ProcessWindowStyle.Hidden;
-
-        return startInfo;
-    }
+﻿using System;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Assignment;
+
+public record struct PingResult(int ExitCode, string? StdOutput);
+
+public class PingProcess
+{
+    private ProcessStartInfo StartInfo { get; } = new("ping");
+
+    public PingResult Run(string hostNameOrAddress)
+    {
+        StartInfo.Arguments = hostNameOrAddress;
+
+        StringBuilder? stringBuilder = null;
+        void updateStdOutput(string? line)
+        {
+            (stringBuilder ??= new StringBuilder()).AppendLine(line);
+        }
+
+        Process process = RunProcessInternal(StartInfo, updateStdOutput, default, default);
+        return new PingResult(process.ExitCode, stringBuilder?.ToString());
+    }
+
+    public Task<PingResult> RunTaskAsync(string hostNameOrAddress)
+    {
+        return Task.Run(() => Run(hostNameOrAddress));
+    }
+
+    public async Task<PingResult> RunAsync(
+        string hostNameOrAddress, CancellationToken cancellationToken = default)
+    {
+        //This feels like extra code, doesn't Task.Run check cancelation token right away and then begin polling IsCancellationRequested after?
+        cancellationToken.ThrowIfCancellationRequested();
+        Task<PingResult> task = Task.Run(() => Run(hostNameOrAddress), cancellationToken);
+        cancellationToken.ThrowIfCancellationRequested();
+        return await task;
+    }
+
+    public async Task<PingResult> RunAsync(params string[] hostNameOrAddresses)
+    {
+        Task<PingResult>[] tasks = hostNameOrAddresses.Select((string x) => RunAsync(x)).ToArray();
+
+        await Task.WhenAll(tasks);
+        return new PingResult(ExitCode: tasks.Max(x => x.Result.ExitCode), StdOutput: string.Join(Environment.NewLine, tasks.Select(outputStringResult => outputStringResult.Result.StdOutput?.Trim() ?? "")));
+
+    }
+    public Task<PingResult> RunLongRunningAsync(
+        string hostNameOrAddress, CancellationToken cancellationToken = default)
+    {
+        return Task.Factory.StartNew(
+            () =>
+        {
+            Task<PingResult> result = RunAsync(hostNameOrAddress);
```

Comment:
Generally if we have a cancellation token, and the method we are calling has an overload with cancellation support, we should use that overload.
```suggestion
            Task<PingResult> result = RunAsync(hostNameOrAddress, cancellationToken );
```